patterns/trackers.md

```diff
@@ -0,0 +1,56 @@
+
+## The Design Solution
+
+Create centralized but redundant "tracker" servers to introduce peers. These trackers have well known addresses so they can be easily found by users, and may even have their addresses hardcoded into client software to aid peer discovery.
+
+Trackers serve as a key/value store, where the "key" is a unique identifier representing content, and the "value" is a list of contact information for introductory peers related to the content. Traditionally this contact information is in the form of an IP address and port number, and may contain additional metadata about the peer's bandwidth and supported functionality. However, the contact information could take a variety of other forms, including onion addresses, public keys, or other content identifiers to indicate related content.
+
+### Examples
+
+Trackers are famously used by torrents to introduce peers to content seeds. They are also used to add new peers to distributed hash tables, or similar communication networks like blockchains. Examples include:
+
+- Torrents
+- Kademlia DHT
+- Blockchains
+- Tor Directory Servers
+- IPFS Bootstrap Peers
+- Cabal
+
+## Why Choose Trackers?
+
+Trackers are a simple technique to solve "first introduction" problems in peer-to-peer systems, provide a natural solution to NAT traversal, and can serve as a bootstrapping step to add clients to a more decentralized network like a distributed hash table.
+
+## Best Practice: How to Implement Trackers
+
+- Include multiple redundant trackers in content links (such as torrent files or magnet links) or hardcoded in client software
+- Have clients periodically re-submit their contact information to trackers to indicate continual interest in content
+- Trackers may prune contact information after an age threshold, or their may include the last "check-in date" with any contact information to allow clients to make pruning decisions on their own
+
+## Potential Problems with Trackers
+
+- Centralized trackers make surveillance easier and introduce a small number of critical hubs that can be disrupted to prevent new clients from joining a network
+- By default, trackers can read any contact information supplied to them, and so can identify the number of peers interested in a content identifier and their contact information
+    - Mitigate this vulnerability by encrypting the contact information before submitting it to a tracker, and using a derived hash of the encryption key as a tracker content identifier, so interested peers will be able to read the peer contact data
+- Trackers only solve discovery of peers based on a shared content identifier, but do _not_ facilitate discovery of new relevant content identifiers
+
+## The Take Away
+
+Trackers are a common component of many peer-to-peer networks, and facilitate peer introduction based on a content identifier. Content identifiers must be shared beforehand, so trackers do not completely solve content discovery.
```

Comment:
Overall your work looks good! I think it could be merged as-is, but here are a couple ideas for improvement if you have time. If not, I can go through and add a few of these changes next week.

I think the content identifier piece could be abstracted out a bit to also be applicable for other kinds of identifiers -- for example, one could use a "pre-shared key" which is based on a particular topic rather than to a particular piece of content.

Reply:
Yes! Especially if the "content" is peer discovery rather than a file. In these circumstances the identifier could refer to a topic, a community, a conceptual "server" (thinking of Cabal here, which acts much like Slack or Discord but the 'server' is internally a collection of p2p feeds), and maybe much more!